### PR TITLE
refactor(common): implement `Datum::[from|to]_protobuf`

### PR DESCRIPTION
**Pull request by BugenZhao**
Signed-off-by: Bugen Zhao <i@bugenzhao.com>I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

To avoid writing serialization/deserialization logic explicitly. 💦

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/common/src/util/memcmp_encoding.rs [hunk 1]
```diff
@@ -430,14 +430,14 @@ mod tests {
     fn test_memcomparable_structs() {
         // NOTE: `NULL`s inside composite type values are always the largest.
 
-        let struct_none = None;
-        let struct_1 = Some(
+        let struct_none = Datum::None;
+        let struct_1 = Datum::Some(
```
Comment (xxchan) [inline src/common/src/util/memcmp_encoding.rs:434]: Why bother change this?

**Review by xxchan — APPROVED** [review]

File: src/common/src/util/memcmp_encoding.rs (hunk 1 above)
Comment (BugenZhao) [reply src/common/src/util/memcmp_encoding.rs:434]: Due to `auto_impl(&)`, it seems the type inference will fail as we use `.into()` in `Some(..)`. 🤣

Comment (stdrc) [reply src/common/src/util/memcmp_encoding.rs:434]: Actually I think `Datum::None`/`Datum::Some` can be a good practice, although changing one place doesn't make big difference in regard to maintainability.

**Review by stdrc — APPROVED** [review]
LGTM

**Pull request merged**